web/api/tests/web_api.c

```diff
@@ -126,43 +220,238 @@ static void build_request(struct web_buffer *wb, const char *url, bool use_cr, s
     buffer_strcat(wb, "\n");
 }
 
-static struct web_client *pre_test_setup()
+/* Note: this is not a CMocka group_test_setup/teardown pair. This is performed per-test.
+*/
+static struct web_client *setup_fresh_web_client()
 {
-    localhost = malloc(sizeof(RRDHOST));
     struct web_client *w = (struct web_client *)malloc(sizeof(struct web_client));
+    memset(w, 0, sizeof(struct web_client));
     w->response.data = buffer_create(NETDATA_WEB_RESPONSE_INITIAL_SIZE);
+    w->response.data->date = 0; // Valgrind uninitialised value
+    w->response.data->expires = 0; // Valgrind uninitialised value
+    w->response.data->options = 0; // Valgrind uninitialised value
     w->response.header = buffer_create(NETDATA_WEB_RESPONSE_HEADER_SIZE);
     w->response.header_output = buffer_create(NETDATA_WEB_RESPONSE_HEADER_SIZE);
     strcpy(w->origin, "*"); // Simulate web_client_create_on_fd()
     w->cookie1[0] = 0;      // Simulate web_client_create_on_fd()
     w->cookie2[0] = 0;      // Simulate web_client_create_on_fd()
+    w->acl = 0x1f;          // Everything on
     return w;
 }
 
-static void post_test_cleanup(struct web_client *w)
+static void destroy_web_client(struct web_client *w)
 {
     buffer_free(w->response.data);
     buffer_free(w->response.header);
     buffer_free(w->response.header_output);
     free(w);
-    free(localhost);
 }
 
+// ---------------------------------- Parameterized test-families -----------------------------------------------------
+// There is no way to pass a parameter block into the setup fixture, we would have to patch CMocka and maintain it
+// locally. (The void **current_state in _run_group_tests would be set from a parameter). This is unfortunate as a
+// parameteric unit-tester needs to be to pass parameters to the fixtures. We are faking this by calculating the
+// space of tests in the launcher, passing an array of identical unit-tests to CMocka and then counting through the
+// parameters in the shared state passed between tests. To initialise this counter structure we use this global to
+// pass from the launcher (test-builder) to the setup-fixture.
+
+void *shared_test_state = NULL;
+
+// -------------------------------- Test family for /api/v1/info ------------------------------------------------------
+
+struct test_def {
+    size_t num_headers; // Index coordinate
+    size_t prefix_len;  // Index coordinate
+    char name[80];
+    size_t full_len;
+    struct web_client *instance; // Used within this single test
+    bool completed;
+    struct test_def *next, *prev;
+};
+
 static void api_info(void **state)
 {
+    (void)state;
+    struct test_def *def = (struct test_def *)shared_test_state;
+    shared_test_state = def->next;
+
+    if (def->prev != NULL && !def->prev->completed && strlen(log_buffer) > 0) {
+        printf("Log of failing case %s:\n", def->prev->name);
```

Comment:
It is printed in the next test. Is it possible to make it appear in the failed test itself?

Reply:
The control-flow in CMocka does not allow that. When the test fails control does not come back to the unit test.